rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,607 @@
+
+[^1]: https://goteleport.com/docs/enroll-resources/database-access/guides/ha/#combined-replicas
+
+## Why
+
+1. Fix random user connection failures caused by network issues in HA agent deployments. This issue is a frequent source of support tickets and user frustration.
+2. Improve database resource onboarding UX
+3. Surface agent->resource connectivity issues early
+4. Improve troubleshooting
+5. Lay the groundwork for collecting agent->resource latency measurements that can be used to make routing decisions based on latency
+
+Related issues:
+- [Database Access upstream DB health checks](https://github.com/gravitational/teleport/issues/20544)
+- [Desktop Access: support latency-based routing to a desktop service](https://github.com/gravitational/teleport/issues/40905)
+- [Detect and report latency measurements for Desktop sessions in the UI](https://github.com/gravitational/teleport/issues/35691)
+
+## Details
+
+Teleport database agents will periodically perform health checks to a database's endpoint and report the results as a health status in the agent's `db_server` heartbeat.
+
+When the Teleport proxy routes a user connection to a database agent it will prioritize database agents that have healthy network connectivity to the database endpoint over agents that do not.
+
+Currently, in an HA DB agent configuration, the proxy will randomly choose one of the DB agents to handle a user connection.
+
+If a subset of the DB agents can't reach the database endpoint, then user connection attempts will randomly fail.
+
+To fix this, the Teleport proxy service will group DB agents by target health status, shuffle each group, and then preferentially dial agents that have healthy network connectivity to the database.
+
+### UX
+
+The web UI will be updated to leverage database health status information.
+
+Unhealthy databases will be displayed with a warning icon or tooltip icon indicating that the database is unhealthy.
+
+Recall that there may be multiple `db_server` heartbeats for the same database, but the web UI only shows a single resource tile for the database.
+If this is the case, then the tooltip will be shown if any of the heartbeats have an unhealthy `target_health.status`.
+The tooltip content may look something like this (or equivalent singular form verbiage if there's only one agent):
+
+    M out of N Teleport database services proxying access to this database cannot reach the database endpoint.
+    <link to healthcheck/network troubleshooting doc>
+
+    Affected Teleport database services:
+    - Hostname: <hostname>
+      UUID: <uuid>
+      Error: <last health check error message>
+    - Hostname: <hostname>
+      UUID: <uuid>
+      Error: <last health check error message>
+
+Databases with a mix of `healthy` and `""` (unknown) status will also show a tooltip that encourages the user to enable health checks or update to a supported version on all of the agents that proxy the db.
```

Comment:
```suggestion
Databases with a mix of `healthy` and `""` (unknown) status will also show a tooltip that encourages the user to update to a supported version on all of the agents that proxy the db.
```

same for example below that with label-based approach there won't mix of disabled vs enabled agent.

Reply:
> You're right that the most common case is there will not be a mix, but there can be a mix in case of dynamic labels and (I think) multiple static databases with the same name but different labels

lol, that sounds like a really bad setup. imagine RBAC decision is on one of those "dynamic" labels and values are different across the HA agents...